### PR TITLE
[SPARK-17032][SQL] Add test cases for methods in ParserUtils.

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

Currently methods in `ParserUtils` are tested indirectly, we should add test cases in `ParserUtilsSuite` to verify their integrity directly.
## How was this patch tested?

New test cases in `ParserUtilsSuite`

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
(Maybe you meant `SPARK-17032` in the title)

**Comment by jiangxb1987** [comment]
@HyukjinKwon Thank you for checking!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala [hunk 1]
```diff
@@ -61,5 +88,39 @@ class ParserUtilsSuite extends SparkFunSuite {
...
+    assert(remainder(showDbsContext) == "")
+
+    assert(remainder(setConfContext.SET.getSymbol) == " example.setting.name=setting.value")
+    assert(remainder(showFuncContext.FUNCTIONS.getSymbol) == " foo.bar")
+    assert(remainder(descFuncContext.EXTENDED.getSymbol) == " bar")
+    assert(remainder(showDbsContext.LIKE.getSymbol) == " 'identifier_with_wildcards'")
+  }
+
+  test("string") {
+    assert(string(showDbsContext.pattern) == "identifier_with_wildcards")
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala]: Could you also add a test for the other `string(node: TerminalNode)` code path?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala [hunk 2]
```diff
@@ -61,5 +88,39 @@ class ParserUtilsSuite extends SparkFunSuite {
     // scalastyle:on nonascii
   }
 
-  // TODO: Add test cases for other methods in ParserUtils
+  test("command") {
+    assert(command(setConfContext) == "set example.setting.name=setting.value")
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala]: We can merge the both the command(..) functions in ParserUtils. Could you do this as part of this PR?

**Comment by hvanhovell** [comment]
@jiangxb1987 thanks for doing this. I left a few comments. Any chance that you also add tests for the remaining methods?

**Comment by jiangxb1987** [comment]
@hvanhovell Thank you for your comments! Will do it as possible as I can.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala (hunk 2 above)
Comment (jiangxb1987) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala]: I searched and found the function `command(stream: CharStream)` was never used, so I simply removed it.

**Comment by hvanhovell** [comment]
@jiangxb1987 are you also adding tests for the remaining `ParserUtils` methods?

**Comment by jiangxb1987** [comment]
@hvanhovell I'm working on it, will try to add test cases for the remaining methods:
1. `operationNotAllowed(message: String, ctx: ParserRuleContext)`;
2. `checkDuplicateKeys[T](keyPairs: Seq[(String, T)], ctx: ParserRuleContext)`;
3. `assert(f: => Boolean, message: String, ctx: ParserRuleContext)`;
4. `withOrigin[T](ctx: ParserRuleContext)(f: => T)`.

**Comment by jiangxb1987** [comment]
I've added test cases for all the methods in `ParserUtils`, and made two small changes in the `ParserUtils` code:
1. Deleted function `command(stream: CharStream)`, merge both `command` method into `command(ctx: ParserRuleContext)`;
2. In method `position(token: Token)` the input param `token` could be `null`, so I added branch statement to handle the `token == null` condition.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala [hunk 3]
```diff
@@ -61,5 +102,84 @@ class ParserUtilsSuite extends SparkFunSuite {
...
+
+    val e = intercept[ParseException] {
+      ParserUtils.assert(f1(emptyContext), message, emptyContext)
+    }.getMessage
+    assert(e.contains(message))
+  }
+
+  test("withOrigin") {
+    val ctx = createDbContext.locationSpec
+    val location = withOrigin(ctx) {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala]: You also need to test the actual origin, use `org.apache.spark.sql.catalyst.trees.CurrentOrigin.get()` for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala [hunk 4]
```diff
@@ -74,7 +70,11 @@ object ParserUtils {
 
   /** Get the origin (line and position) of the token. */
   def position(token: Token): Origin = {
-    Origin(Option(token.getLine), Option(token.getCharPositionInLine))
+    if (token != null) {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala]: how about:
```scala
val opt = Option(token)
Origin(opt.map(_.getLine), opt.map(_.getCharPositionInLine))
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala [hunk 5]
```diff
@@ -61,5 +102,87 @@ class ParserUtilsSuite extends SparkFunSuite {
...
+    val e = intercept[ParseException] {
+      ParserUtils.assert(f1(emptyContext), message, emptyContext)
+    }.getMessage
+    assert(e.contains(message))
+  }
+
+  test("withOrigin") {
+    val ctx = createDbContext.locationSpec
+    val current = CurrentOrigin.get
+    val location = withOrigin(ctx) {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala]: you also need to check if the origin of location to make this a complete tests, e.g.:
```scala
val (location, origin) = withOrigin(ctx) {
  (string(ctx.string), CurrentOrigin.get)
}
assert(origin, Origin(Some(2), Some(37)))
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala [hunk 6]
```diff
@@ -61,5 +102,88 @@ class ParserUtilsSuite extends SparkFunSuite {
...
+
+  test("position") {
+    assert(position(setConfContext.start) == Origin(Some(1), Some(0)))
+    assert(position(showFuncContext.stop) == Origin(Some(1), Some(19)))
+    assert(position(descFuncContext.describeFuncName.start) == Origin(Some(1), Some(27)))
+    assert(position(createDbContext.locationSpec.start) == Origin(Some(3), Some(27)))
+    assert(position(emptyContext.stop) == Origin(None, None))
+  }
+
+  test("assert") {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala]: We renamed assert into validate. So you will have to update. Sorry about that.

**Comment by hvanhovell** [comment]
LGTM - pending Jenkins.

**Comment by hvanhovell** [comment]
Merging to master. Thanks!

**Comment by jiangxb1987** [comment]
@hvanhovell Thank you for your help!

**Pull request closed without merging**